File: android/app/src/main/java/io/github/yurisismotto/anyflow/notifications/NotificationPolicy.kt
package io.github.yurisismotto.anyflow.notifications

import org.json.JSONArray
import org.json.JSONObject

/**
 * How much of a notification crosses the wire while the source device is
 * locked.
 *
 * The reduction happens **at the source, before encoding**, so content
 * withheld by policy never enters the protobuf, never touches the socket and
 * never exists on the desktop. That is the strong form of a privacy control:
 * there is nothing to leak because there is nothing there.
 *
 * Unlocking is **not retroactive**. Content withheld while locked is not
 * delivered later, because holding it somewhere until the phone unlocks would
 * be a notification history by another name.
 */
enum class LockPolicy {
    /** Everything, as if unlocked. */
    FULL,

    /** `app_label` only. `title` and `body` are empty and `redacted` is set. */
    APP_ONLY,

    /** Nothing is sent at all. */
    SUPPRESS,
    ;

    companion object {
        /** ADR-0015 §7: the default, on both ends. */
        val DEFAULT = APP_ONLY

        /** An unrecognised stored value is the default, never [FULL]. */
        fun fromName(name: String?): LockPolicy =
            entries.firstOrNull { it.name == name } ?: DEFAULT
    }
}

/**
 * What this phone will mirror to one computer.
 *
 * ## Two different questions, again
 *
 * Exactly as `clipboard.v1` (see [io.github.yurisismotto.anyflow.clipboard.ClipboardPolicy]):
 *
 *  * the **grant** — in the trust store, beside `files.v1`'s — answers "may
 *    this computer receive notifications from me at all?", and it is never
 *    automatic and never in `auto_grant` (ADR-0015 §4);
 *  * the **policy** — this type — answers "which of them, and how much of
 *    each?".
 *
 * And there is a third question neither of them answers: whether Android has
 * given this app notification access at all. Collapsing any two of the three
 * would mean a person who wanted a phone-side capability had silently
 * authorised a network destination (ADR-0017 §6).
 *
 * ## A peer can never set its own policy
 *
 * Every field here is decided and stored locally. **There is no protocol
 * message that writes any of them**, and that is enforced by the absence of a
 * field in the schema rather than by a check that a later change could invert.
 *
 * ## Deny by default
 *
 * [allowedApps] starts **empty**, so a freshly granted computer receives
 * nothing until a person names an application. That is the single most
 * important default in this capability: on the certification hardware the
 * platform's own OTP redaction did not fire at all (POC-NOTIF-01), which makes
 * the per-app allow-list the only effective control there is — an argument for
 * deny-by-default rather than against it.
 *
 * An app installed after the grant is therefore *also* denied, without needing
 * a rule of its own: it is simply not in the set. N3 owns the picker that
 * fills it, and the passive "3 new apps are not being shared" affordance.
 *
 * ## What is deliberately not here
 *
 * No OTP heuristic, no keyword list, no banking-app detection, no
 * category-based rule. A guess dressed as a security control is worse than an
 * honest boundary, because the user trusts it and it is wrong in cases neither
 * party can predict (THREAT_MODEL T10, ADR-0015 §5).
 */
data class NotificationPolicy(
    /**
     * May this computer receive mirrored notifications at all?
     *
     * Defaults **on**, for the reason `clipboard.v1`'s `allowSend` does:
     * reaching this field already required an explicit, revocable grant, so it
     * is what the person just asked for. It is not a permissive default,
     * because [allowedApps] is empty and nothing is mirrored until it is not.
     */
    val allowMirror: Boolean = true,
    /**
     * The packages this computer may receive. **Empty by default.**
     *
     * AnyFlow's own package is not in this set and could not help if it were:
     * the own-package rule is applied before the filter is consulted and there
     * is no setting that turns it off.
     */
    val allowedApps: Set<String> = emptySet(),
    /**
     * Work-profile notifications, behind their own switch and **off**.
     *
     * A listener in the personal profile receives work-profile notifications
     * unless the administrator blocks it. Mirroring an employer's data onto a
     * personal machine is not a decision AnyFlow makes on someone's behalf,
     * and a user who shares "Slack" from their personal profile has not
     * thereby asked to share work Slack.
     */
    val includeWorkProfile: Boolean = false,
    /**
     * Ongoing and foreground-service notifications. **Off.**
     *
     * "Maps is navigating", "Spotify is playing". They are not events, they
     * update constantly, they are usually not dismissible, and mirroring them
     * fills a desktop with rows that cannot be cleared.
     */
    val includeOngoing: Boolean = false,
    /** ADR-0015 §7: what crosses the wire while this phone is locked. */
    val whenSourceLocked: LockPolicy = LockPolicy.DEFAULT,
    /**
     * Whether a desktop dismissal may cancel the notification here.
     *
     * **Off by default** (ADR-0015 §6), and **not implemented in N1**: this
     * wave contains no path from an inbound message to `cancelNotification`,
     * and Android announces no `DISMISS_TARGET` role. The field is stored so
     * N4 has somewhere to read the answer from, and so the default it must
     * respect is already written down.
     */
    val allowDismissSync: Boolean = false,
) {

    /**
     * Whether [packageName] may be mirrored to this computer.
     *
     * Deny by default: an app that has never been named is not shared, and
     * neither is one installed after the grant.
     */
    fun allowsApp(packageName: String): Boolean =
        allowMirror && packageName in allowedApps

    /** Contains no content and no identity; safe to log. */
    fun describe(): String =
        "mirror=${onOff(allowMirror)} apps=${allowedApps.size} " +
            "work-profile=${onOff(includeWorkProfile)} ongoing=${onOff(includeOngoing)} " +
            "locked=$whenSourceLocked dismiss-sync=${onOff(allowDismissSync)}"

    /**
     * Stores the settings, and only the settings.
     *
     * Package names are settings — the user chose them — and no notification
     * content reaches this object or the file it is written to.
     */
    fun toJson(): JSONObject = JSONObject().apply {
        put(KEY_ALLOW_MIRROR, allowMirror)
        put(KEY_ALLOWED_APPS, JSONArray(allowedApps.sorted()))
        put(KEY_INCLUDE_WORK_PROFILE, includeWorkProfile)
        put(KEY_INCLUDE_ONGOING, includeOngoing)
        put(KEY_WHEN_SOURCE_LOCKED, whenSourceLocked.name)
        put(KEY_ALLOW_DISMISS_SYNC, allowDismissSync)
    }

    companion object {
        /** Nothing is permitted, whatever the stored flags say. */
        val DENIED = NotificationPolicy(
            allowMirror = false,
            allowedApps = emptySet(),
            includeWorkProfile = false,
            includeOngoing = false,
            whenSourceLocked = LockPolicy.SUPPRESS,
            allowDismissSync = false,
        )

        private const val KEY_ALLOW_MIRROR = "allowMirror"
        private const val KEY_ALLOWED_APPS = "allowedApps"
        private const val KEY_INCLUDE_WORK_PROFILE = "includeWorkProfile"
        private const val KEY_INCLUDE_ONGOING = "includeOngoing"
        private const val KEY_WHEN_SOURCE_LOCKED = "whenSourceLocked"
        private const val KEY_ALLOW_DISMISS_SYNC = "allowDismissSync"

        /**
         * Reads a stored policy.
         *
         * A trust store written before this capability existed has no policy
         * object at all, and every field falls back to its documented default
         * rather than to `false` — or, for the app list, to "everything". An
         * upgrade must not silently widen what is shared.
         */
        fun fromJson(json: JSONObject?): NotificationPolicy {
            if (json == null) return NotificationPolicy()
            val apps = json.optJSONArray(KEY_ALLOWED_APPS)
            return NotificationPolicy(
                allowMirror = json.optBoolean(KEY_ALLOW_MIRROR, true),
                allowedApps = buildSet {
                    if (apps != null) {
                        for (index in 0 until apps.length()) {
                            apps.optString(index).takeIf { it.isNotEmpty() }?.let { add(it) }
                        }
                    }
                },
                includeWorkProfile = json.optBoolean(KEY_INCLUDE_WORK_PROFILE, false),
                includeOngoing = json.optBoolean(KEY_INCLUDE_ONGOING, false),
                whenSourceLocked = LockPolicy.fromName(
                    json.optString(KEY_WHEN_SOURCE_LOCKED, LockPolicy.DEFAULT.name),
                ),
                allowDismissSync = json.optBoolean(KEY_ALLOW_DISMISS_SYNC, false),
            )
        }

        private fun onOff(value: Boolean) = if (value) "on" else "off"
    }
}

File: desktop/core/src/notification_policy.rs
//! Per-peer `notifications.v1` policy, as this device stores it.
//!
//! This type lives in `anyflow-core` rather than in the notifications
//! capability for the reason [`crate::clipboard_policy`] gives: it is
//! *persisted*, and the trust store is core's. A capability crate cannot own
//! a field of [`crate::store::TrustedPeer`] without core depending on it, and
//! core depending on a capability would invert the plugin model. The
//! capability re-exports it, so callers still say
//! `anyflow_capability_notifications::NotificationPolicy`.
//!
//! # Two different questions, and a third
//!
//! * The **grant** — in the trust store beside `files.v1`'s and
//!   `clipboard.v1`'s — answers *"may this device speak notifications with me
//!   at all?"*. It is never in `auto_grant` (ADR-0008, ADR-0015 §4).
//! * The **policy** — this type — answers *"and how much of each, and when?"*.
//! * The **roles** each side announces answer *"what can that device
//!   physically do right now?"*, and are never an authorization input
//!   (ADR-0017 §6).
//!
//! Collapsing any two of the three would mean somebody who wanted one thing
//! had silently authorised another.
//!
//! # A peer can never set its own policy
//!
//! Every field here is decided locally and stored locally. **There is no
//! protocol message that writes any of them** — `notifications_v1.proto` has
//! six bodies and none of them carries a setting — so the rule is an absence
//! rather than a check a later change could invert. That is NOTIF-SEC-20, and
//! it is the same guarantee `clipboard.v1` makes.
//!
//! # What is deliberately not here
//!
//! The **source-side** half of the policy — the per-app allow-list, the
//! work-profile switch, the ongoing switch, `when_source_locked` — belongs to
//! the device that owns the notifications, which in v1 is Android, and it is
//! stored there (`NotificationPolicy.kt`). A desktop copy of those fields
//! would be a second place to look and a second place to be wrong: the sink
//! cannot ask for more than it is sent, so the fields would have no effect
//! here even if they were set. N3 owns the surface that edits the Android
//! side; N2 stores only what this machine can actually enforce.

use serde::{Deserialize, Serialize};

/// How much of a notification is presented while **this** device is locked.
///
/// The names match the source-side enum deliberately: the two policies
/// compose, the source's applies first because it decides what crosses the
/// wire, and a reader comparing the two ends should not have to translate.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum LockPolicy {
    /// Everything, as if unlocked.
    Full,
    /// The application's name only. No title, no body.
    #[default]
    AppOnly,
    /// Nothing is displayed at all, and an existing mirror is closed.
    Suppress,
}

impl LockPolicy {
    /// ADR-0015 §7: the default, on both ends.
    pub const DEFAULT: Self = Self::AppOnly;

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Full => "full",
            Self::AppOnly => "app-only",
            Self::Suppress => "suppress",
        }
    }

    /// Parses the spelling the CLI uses. Unknown input is **not** guessed at:
    /// a typo must not silently select the most permissive option.
    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "full" => Some(Self::Full),
            "app-only" | "app_only" | "apponly" => Some(Self::AppOnly),
            "suppress" | "none" => Some(Self::Suppress),
            _ => None,
        }
    }
}

/// What this device will display for one peer.
///
/// # Defaults
///
/// ```text
/// allow_mirror       true      what the grant was for
/// when_sink_locked   AppOnly   ADR-0015 §7
/// allow_dismiss_sync false     ADR-0015 §6 — the only outbound authority
/// ```
///
/// `allow_mirror` defaulting on is not a silent widening: reaching this type
/// at all already required an explicit `anyflow grant <device>
/// notifications.v1`, which is never automatic. What the default *does* say is
/// that the grant means what it looks like it means.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct NotificationPolicy {
    /// May this peer's notifications be displayed here at all?
    ///
    /// Off answers every upsert `REJECTED_POLICY` and closes any mirror the
    /// peer already has.
    #[serde(default = "default_true")]
    pub allow_mirror: bool,

    /// What is displayed while this desktop session is locked.
    ///
    /// The reduction happens **before** the notification is handed to the
    /// platform, so what policy withholds is never given to the notification
    /// server at all. A D-Bus client cannot tell GNOME how to render on the
    /// lock screen; the only lever is what goes into the call.
    #[serde(default)]
    pub when_sink_locked: LockPolicy,

    /// Whether a human closing a mirror here may dismiss the notification on
    /// the source device.
    ///
    /// **Off by default** (ADR-0015 §6), and per peer. Every other field in
    /// this type decides what this desktop will *display*; this one is the only
    /// authority that points outward, and it is the reason it is asked for
    /// separately even though the action it authorises is small.
    ///
    /// It is one of four independent gates on a dismissal, and the other three
    /// are deliberately not here: the peer's own `DISMISS_TARGET` claim, this
    /// device's `DISMISS_REPORTER` role, and the source's own
    /// `allowDismissSync`. A policy cannot manufacture a capability, and a
    /// capability cannot manufacture a permission.
    #[serde(default)]
    pub allow_dismiss_sync: bool,
}

fn default_true() -> bool {
    true
}

impl Default for NotificationPolicy {
    fn default() -> Self {
        Self {
            allow_mirror: true,
            when_sink_locked: LockPolicy::DEFAULT,
            allow_dismiss_sync: false,
        }
    }
}

impl NotificationPolicy {
    /// The policy for a peer that has no `notifications.v1` grant: nothing is
    /// permitted, whatever the stored flags say.
    ///
    /// Returned by the authorizer so that a caller cannot ask "what is the
    /// policy?" and forget "is it granted?". Every field here is already
    /// closed, and `when_sink_locked` is `Suppress` rather than the default so
    /// that even a caller that read only that field reaches the safe answer.
    pub const DENIED: Self = Self {
        allow_mirror: false,
        when_sink_locked: LockPolicy::Suppress,
        allow_dismiss_sync: false,
    };

    /// True when nothing at all is permitted.
    pub fn is_denied(&self) -> bool {
        !self.allow_mirror
    }

    /// Whether a dismissal here may travel to the source.
    ///
    /// `allow_dismiss_sync` alone is not enough: a policy with `allow_mirror`
    /// off and `allow_dismiss_sync` on is contradictory and resolves to "no",
    /// the same containment rule `ClipboardPolicy::may_auto_send` applies.
    pub fn may_sync_dismissals(&self) -> bool {
        self.allow_mirror && self.allow_dismiss_sync
    }

    /// Renders the policy as the flag names the CLI uses. Contains no
    /// content and no identity; safe to log.
    pub fn describe(&self) -> String {
        let on = |b: bool| if b { "on" } else { "off" };
        format!(
            "mirror={} locked={} dismiss-sync={}",
            on(self.allow_mirror),
            self.when_sink_locked.as_str(),
            on(self.allow_dismiss_sync),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_defaults_are_the_documented_ones() {
        let p = NotificationPolicy::default();
        assert!(p.allow_mirror, "a granted peer's notifications are shown");
        assert_eq!(
            p.when_sink_locked,
            LockPolicy::AppOnly,
            "a locked desktop shows the app name and nothing else"
        );
        assert!(!p.allow_dismiss_sync, "dismiss sync must never default on");
        assert!(!p.may_sync_dismissals());
    }

    #[test]
    fn denied_closes_everything_including_the_lock_policy() {
        let p = NotificationPolicy::DENIED;
        assert!(p.is_denied());
        assert!(!p.may_sync_dismissals());
        assert_eq!(p.when_sink_locked, LockPolicy::Suppress);
    }

    #[test]
    fn withdrawing_mirroring_defeats_a_stale_dismiss_flag() {
        let p = NotificationPolicy {
            allow_mirror: false,
            allow_dismiss_sync: true,
            ..NotificationPolicy::default()
        };
        assert!(!p.may_sync_dismissals());
    }

    #[test]
    fn missing_fields_deserialize_to_the_documented_defaults() {
        // A trust store written before this capability existed has no policy
        // object at all, and one written by an older build may be missing
        // individual flags. Neither may silently widen what is displayed.
        let p: NotificationPolicy = serde_json::from_str("{}").expect("empty object is valid");
        assert_eq!(p, NotificationPolicy::default());

        let p: NotificationPolicy = serde_json::from_str(r#"{"allow_dismiss_sync":true}"#)
            .expect("partial object is valid");
        assert!(p.allow_mirror);
        assert_eq!(p.when_sink_locked, LockPolicy::AppOnly);
        assert!(p.allow_dismiss_sync);
    }

    #[test]
    fn an_unknown_lock_policy_spelling_is_refused_rather_than_guessed() {
        assert_eq!(LockPolicy::parse("full"), Some(LockPolicy::Full));
        assert_eq!(LockPolicy::parse("app-only"), Some(LockPolicy::AppOnly));
        assert_eq!(LockPolicy::parse("SUPPRESS"), Some(LockPolicy::Suppress));
        // Not `Full`, and not the default either: the caller is told.
        assert_eq!(LockPolicy::parse("ful"), None);
        assert_eq!(LockPolicy::parse(""), None);
    }

    #[test]
    fn describe_is_stable_and_carries_no_content() {
        assert_eq!(
            NotificationPolicy::default().describe(),
            "mirror=on locked=app-only dismiss-sync=off"
        );
    }
}
